x/auth/ante.go

```diff
@@ -256,23 +257,33 @@ func adjustFeesByGas(fees sdk.Coins, gas int64) sdk.Coins {
 	return fees.Plus(gasFees)
 }
 
-// Deduct the fee from the account.
-// We could use the CoinKeeper (in addition to the AccountKeeper,
-// because the CoinKeeper doesn't give us accounts), but it seems easier to do this.
-func deductFees(acc Account, fee StdFee) (Account, sdk.Result) {
-	coins := acc.GetCoins()
-	feeAmount := fee.Amount
+// deductFees attempts to deduct fees from a given address. Upon success the
+// updated account and a result is returned.
+func deductFees(ctx sdk.Context, acc Account, fee sdk.Coins) (Account, sdk.Result) {
+	// TODO: Do we need charge gas for fee deduction?
+	oldCoins := acc.GetCoins()
```

Comment:
Hmm so I can make a `SubtractCoinsAndSave` *auxiliary function*  in `account.go` that performs this. However, it seems it will only be used in the ante handler (as of now) even though the logic is nearly identical in `subtractCoins` in the bank keeper. The reason is the bank keeper is slightly different in that it charges gas for many ops and uses the acc keeper under the hood (abstracted away) with no direct access to an account.

Do you have any exact suggestions?

Reply:
Actually, I may know a way of doing this.

Reply:
I see, you're right that as is it would not be possible to use the exact same function in both places. 

Mostly, I think it would be less easy to misuse in callers if we could write this logic - and the logic for `DelegateCoins` and `UndelegateCoins` - in a way that *didn't* require type-switching on `VestingAccount` but was instead a method on the account interface (so the vesting account could do one thing and the base account another). Do you think that would be possible?

Reply:
Ok, updated. I essentially moved a few methods to the parent `Account` interface and this allowed me to cleanup and remove type checking in the bank keeper.